File: nodejs/lesson03/solution/hello.js
'use strict';

const express = require('express')
const app = express()
const initTracer = require('../../lib/tracing').initTracer;
var request = require('request-promise');

const { Tags, FORMAT_HTTP_HEADERS } = require('opentracing');

const tracer = initTracer('hello-world');

const port = 8080;

app.listen(port, function () {
    console.log('Hello app listening on port ' + port);
})

// main endpoint
app.get('/:str', main);

function main(req, res) {
    const input = req.params.str;

    const span = tracer.startSpan('say-hello-request');

    span.log({
        'event': 'sayHelloRequest',
        'value': input
    });

    format_string(input, span)
        .then( data => {
            return print_hello(data, span);
        })
        .then( data => {
            span.setTag(Tags.HTTP_STATUS_CODE, 200)
            span.finish()
            res.send('Got response back from service: ' + data);
        })
        .catch( err => {
            console.error(err.message);
            span.setTag(Tags.ERROR, true)
            span.setTag(Tags.HTTP_STATUS_CODE, err.statusCode || 500);
            span.finish();
            res.send('Sorry, an error happened during call downstream service!');
        });

}

function format_string(input, root_span) {
    const url = 'http://localhost:8081/format/' + input;
    const fn = 'format';

    const span = tracer.startSpan(fn, {childOf: root_span.context()});
    span.log({
        'event': 'format-string',
        'value': input
    });

    return http_get(fn, url, span); 
}  

function print_hello(input, root_span) {
    const url = 'http://localhost:8082/publish/' + input;
    const fn = 'publish';

    const span = tracer.startSpan(fn, {childOf: root_span.context()});
    span.log({
        'event': 'print-string',
        'value': input
    });
    return http_get(fn, url, span);
}

function http_get(fn, url, span) {
    const method = 'GET';
    const headers = {};
    
    span.setTag(Tags.HTTP_URL, url);
    span.setTag(Tags.HTTP_METHOD, method);
    span.setTag(Tags.SPAN_KIND, Tags.SPAN_KIND_RPC_CLIENT);
    // Send span context via request headers (parent id etc.)
    tracer.inject(span, FORMAT_HTTP_HEADERS, headers);

    return request({url, method, headers})
            .then( data => {
                span.finish();
                return data;
            });

}

  

File: java/src/main/java/lesson03/solution/Hello.java
package lesson03.solution;

import io.opentracing.Scope;
import io.opentracing.propagation.Format.Builtin;
import io.opentracing.tag.Tags;

import java.io.IOException;
import java.util.Iterator;
import java.util.Map.Entry;
import com.google.common.collect.ImmutableMap;
import com.uber.jaeger.Tracer;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.HttpUrl;
import lib.Tracing;

public class Hello {

    private final Tracer tracer;
    private final OkHttpClient client;

    private Hello(Tracer tracer) {
        this.tracer = tracer;
        this.client = new OkHttpClient();
    }

    private static class RequestBuilderCarrier implements io.opentracing.propagation.TextMap {
        private final Request.Builder builder;

        RequestBuilderCarrier(Request.Builder builder) {
            this.builder = builder;
        }

		@Override
		public Iterator<Entry<String, String>> iterator() {
			throw new UnsupportedOperationException("carrier is write-only");
		}

		@Override
		public void put(String key, String value) {
			builder.addHeader(key, value);
		}

    }

    private String getHttp(int port, String path, String param, String value) {
        try {
            HttpUrl url = new HttpUrl.Builder().scheme("http").host("localhost").port(port).addPathSegment(path)
                    .addQueryParameter(param, value).build();
            Request.Builder requestBuilder = new Request.Builder().url(url);

            Tags.SPAN_KIND.set(tracer.activeSpan(), Tags.SPAN_KIND_CLIENT);
            Tags.HTTP_METHOD.set(tracer.activeSpan(), "GET");
            Tags.HTTP_URL.set(tracer.activeSpan(), url.toString());
            tracer.inject(tracer.activeSpan().context(), Builtin.HTTP_HEADERS, new RequestBuilderCarrier(requestBuilder));

            Request request = requestBuilder.build();
            Response response = client.newCall(request).execute();
            if (response.code() != 200) {
                throw new RuntimeException("Bad HTTP result: " + response);
            }
            return response.body().string();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private void sayHello(String helloTo) {
        try (Scope scope = tracer.buildSpan("say-hello").startActive(true)) {
            scope.span().setTag("hello-to", helloTo);

            String helloStr = formatString(helloTo);
            printHello(helloStr);
        }
    }

    private String formatString(String helloTo) {
        try (Scope scope = tracer.buildSpan("formatString").startActive(true)) {
            String helloStr = getHttp(8081, "format", "helloTo", helloTo);
            scope.span().log(ImmutableMap.of("event", "string-format", "value", helloStr));
            return helloStr;
        }
    }

    private void printHello(String helloStr) {
        try (Scope scope = tracer.buildSpan("printHello").startActive(true)) {
            getHttp(8082, "publish", "helloStr", helloStr);
            scope.span().log(ImmutableMap.of("event", "println"));
        }
    }

    public static void main(String[] args) {
        if (args.length != 1) {
            throw new IllegalArgumentException("Expecting one argument");
        }
        String helloTo = args[0];
        Tracer tracer = Tracing.init("hello-world");
        new Hello(tracer).sayHello(helloTo);
        tracer.close();
        System.exit(0); // okhttpclient sometimes hangs maven otherwise
    }
}
